### PR TITLE
JsonObject cannot map to class with Instant or byte array field

### LINKED ISSUE #2753: JsonObject cannot map to class with Instant or byte array field
The Jackson mapper is configured to serialize `Instant` and `byte` arrays properly (required for `io.vertx.core.json.JsonObject#mapFrom` method).

But it's not configured to deserialize properly (in method `io.vertx.core.json.JsonObject#mapTo`).

### PR DESCRIPTION
**Pull request by tsegismont**
Fixes #2753

### CODE REVIEW

**Comment by pmlopes** [comment]
@tsegismont sorry, but I think this is more or less the same as https://github.com/eclipse-vertx/vert.x/pull/2507 can you verify?

**Comment by tsegismont** [comment]
@pmlopes it looks similar indeed. Do you want me to close this one and merge the work in the other PR?

I'm not sure why the other one was not merged. I read concerns about deserializing to `JsonObject`, but that wouldn't happen, as the deserializers will only be used when trying to map to a custom class, not `JsonObject` or `JsonArray`.

**Comment by tsegismont** [comment]
I've incorporated @EmadAlblueshi 's work (thanks again buddy!). So we have now something consistent for both `Instant` and byte arrays.

@pmlopes Can you please review? Thanks!

File: src/main/java/io/vertx/core/json/Json.java [hunk 1]
```diff
@@ -239,12 +243,36 @@ public void serialize(Instant value, JsonGenerator jgen, SerializerProvider prov
...
+
+  private static class ByteArrayDeserializer extends JsonDeserializer<byte[]> {
+
+    @Override
+    public byte[] deserialize(JsonParser p, DeserializationContext ctxt) throws IOException, JsonProcessingException {
+      String text = p.getText();
+      try {
+        return Base64.getDecoder().decode(text);
+      } catch (IllegalArgumentException e) {
+        throw new InvalidFormatException(p, "Expected a base64 encoded byte array", text, Instant.class);
```
Comment (EmadAlblueshi) [inline src/main/java/io/vertx/core/json/Json.java]: Why Instant.class ?

**Comment by EmadAlblueshi** [comment]
Thank you @tsegismont I really appreciate it :)

File: src/main/java/io/vertx/core/json/Json.java (hunk 1 above)
Comment (tsegismont) [reply src/main/java/io/vertx/core/json/Json.java]: This is what the Jackson `InvalidFormatException` expects. It helps give the user better failure messages.

**Comment by tsegismont** [comment]
@pmlopes fyi, the PR build passed twice (as I was trying to keep @EmadAlblueshi authorship in commits without success and pushed again and again). But just failed for another reason.

File: src/main/java/io/vertx/core/json/Json.java [hunk 2]
```diff
@@ -239,12 +243,36 @@ public void serialize(Instant value, JsonGenerator jgen, SerializerProvider prov
...
+    }
+  }
+
   private static class ByteArraySerializer extends JsonSerializer<byte[]> {
-    private final Base64.Encoder BASE64 = Base64.getEncoder();
 
     @Override
     public void serialize(byte[] value, JsonGenerator jgen, SerializerProvider provider) throws IOException {
-      jgen.writeString(BASE64.encodeToString(value));
+      jgen.writeString(Base64.getEncoder().encodeToString(value));
```
Comment (pmlopes) [inline src/main/java/io/vertx/core/json/Json.java]: This is wrong according to: https://tools.ietf.org/html/rfc7493#section-4.4 it should be: https://docs.oracle.com/javase/8/docs/api/java/util/Base64.html#getUrlEncoder--

File: src/main/java/io/vertx/core/json/Json.java [hunk 3]
```diff
@@ -239,12 +243,36 @@ public void serialize(Instant value, JsonGenerator jgen, SerializerProvider prov
...
+    }
+  }
+
+  private static class ByteArrayDeserializer extends JsonDeserializer<byte[]> {
+
+    @Override
+    public byte[] deserialize(JsonParser p, DeserializationContext ctxt) throws IOException, JsonProcessingException {
+      String text = p.getText();
+      try {
+        return Base64.getDecoder().decode(text);
```
Comment (pmlopes) [inline src/main/java/io/vertx/core/json/Json.java]: Same here it should be the url/filename decoder.

**Comment by tsegismont** [comment]
As discussed off-list with @pmlopes , I am going to merge this PR.

It is true that the RFC recommends to use `base64url` encoding, but this is a breaking change and we want this code to be backported to 3.6.

As a consequence, after this PR merged and the code backported to 3.6 branch, I will create new issues in `core` and `codegen` to switch to `base64url` encoding. I will also update the breaking changes page.

**Comment by EmadAlblueshi** [comment]
Thanks 👍

**Comment by tsegismont** [comment]
You're welcome!

**Pull request merged**